File: android-native/app/src/main/java/com/example/kpkn/domain/training/ProgramCalendarEngine.kt
package com.example.kpkn.domain.training

import com.example.kpkn.data.models.Program
import com.example.kpkn.data.models.ProgramCalendarization
import com.example.kpkn.data.models.ProgramCalendarizationMode
import com.example.kpkn.data.models.ProgramKeyDate
import com.example.kpkn.data.models.ProgramStructure
import com.example.kpkn.data.models.ProgramWeek
import com.example.kpkn.data.models.Session
import com.example.kpkn.data.models.SimpleProgramKind
import com.example.kpkn.data.models.isSimpleProgram
import com.example.kpkn.data.models.resolvedSchedulePlan
import com.example.kpkn.data.models.suggestCalendarTrainingDays
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeParseException
import java.time.temporal.ChronoUnit

enum class ProgramEndDateStatus {
    NONE,
    MATCHES_PROJECTED,
    BEFORE_PROJECTED,
    AFTER_PROJECTED,
    INVALID_MANUAL,
}

data class CalendarWeekProjection(
    val weekId: String,
    val weekName: String,
    val macroIndex: Int,
    val blockIndex: Int,
    val mesoIndex: Int,
    val weekIndex: Int,
    val blockId: String,
    val blockName: String,
    val startDate: LocalDate,
    val endDate: LocalDate,
    val outsideProgramDays: Set<Int> = emptySet(),
    val trainingDayDates: Map<Int, LocalDate> = emptyMap(),
    val keyDates: List<ProgramKeyDate> = emptyList(),
) {
    fun contains(date: LocalDate): Boolean = !date.isBefore(startDate) && !date.isAfter(endDate)
}

data class ProgramCalendarProjection(
    val enabled: Boolean,
    val mode: ProgramCalendarizationMode?,
    val strictStart: Boolean,
    val activatedByCompetition: Boolean,
    val startDate: LocalDate?,
    val projectedEndDate: LocalDate?,
    val manualEndDate: LocalDate?,
    val endDateStatus: ProgramEndDateStatus,
    val weeks: List<CalendarWeekProjection>,
) {
    val effectiveEndDate: LocalDate?
        get() = manualEndDate ?: projectedEndDate

    fun weekForDate(date: LocalDate): CalendarWeekProjection? =
        weeks.firstOrNull { it.contains(date) }

    fun scheduledDateFor(session: Session, weekId: String): LocalDate? {
        val week = weeks.firstOrNull { it.weekId == weekId } ?: return null
        val day = session.dayOfWeek?.coerceIn(1, 7) ?: return week.startDate
        return week.trainingDayDates[day]
    }
}

object ProgramCalendarEngine {
    fun parseIsoDate(raw: String?): LocalDate? {
        if (raw.isNullOrBlank()) return null
        return try {
            LocalDate.parse(raw, DateTimeFormatter.ISO_LOCAL_DATE)
        } catch (_: DateTimeParseException) {
            null
        }
    }

    fun isCalendarized(program: Program): Boolean {
        val calendar = program.calendarization ?: return false
        return when (calendar.mode) {
            ProgramCalendarizationMode.ADVANCED_COMPETITION ->
                !program.isSimpleProgram && !program.timelineStartDate.isNullOrBlank()
            ProgramCalendarizationMode.SIMPLE_DATED ->
                program.isSimpleProgram &&
                    program.simpleProgramKind == SimpleProgramKind.CALENDARIZED &&
                    !program.timelineStartDate.isNullOrBlank()
        }
    }

    fun defaultCompetitionCalendarization(): ProgramCalendarization =
        ProgramCalendarization(
            mode = ProgramCalendarizationMode.ADVANCED_COMPETITION,
            strictStart = true,
            activatedByCompetition = true,
        )

    /** Advanced dated calendar without requiring a competition key date. */
    fun defaultAdvancedDatedCalendarization(): ProgramCalendarization =
        ProgramCalendarization(
            mode = ProgramCalendarizationMode.ADVANCED_COMPETITION,
            strictStart = true,
            activatedByCompetition = false,
        )

    fun defaultSimpleDatedCalendarization(): ProgramCalendarization =
        ProgramCalendarization(
            mode = ProgramCalendarizationMode.SIMPLE_DATED,
            strictStart = false,
            activatedByCompetition = false,
        )

    fun project(program: Program): ProgramCalendarProjection {
        val calendar = program.calendarization
        val plan = program.resolvedSchedulePlan()
        val anchorDate = parseIsoDate(program.timelineStartDate) ?: parseIsoDate(plan.anchorDate)
        if (calendar == null || anchorDate == null) {
            return ProgramCalendarProjection(
                enabled = false,
                mode = calendar?.mode,
                strictStart = calendar?.strictStart == true,
                activatedByCompetition = calendar?.activatedByCompetition == true,
                startDate = anchorDate,
                projectedEndDate = null,
                manualEndDate = parseIsoDate(calendar?.manualEndDate),
                endDateStatus = ProgramEndDateStatus.NONE,
                weeks = emptyList(),
            )
        }
        val start: LocalDate = anchorDate
        var cursor = start
        var globalWeekIndex = 0
        val weeks = mutableListOf<CalendarWeekProjection>()

        program.macrocycles.forEachIndexed { macroIndex, macro ->
            var globalMesoIndex = 0
            macro.blocks.forEachIndexed { blockIndex, block ->
                block.mesocycles.forEach { meso ->
                    val mesoIndex = globalMesoIndex++
                    meso.weeks.forEach { week ->
                        if (week.isLoopWeek) return@forEach
                        val weekStart = cursor
                        val weekEnd = weekStart.plusDays(6)
                        val trainingDays = resolveTrainingDays(program, week)
                        val dayDates = trainingDatesFor(
                            weekStart = weekStart,
                            weekEnd = weekEnd,
                            trainingDays = trainingDays,
                            explicit = week.trainingDayDates,
                        )
                        val marks = program.keyDates.filter { keyDateIntersects(it, weekStart, weekEnd) }
                        val resolvedDays = dayDates.keys
                        val outsideDays = (1..7).filter { day ->
                            day !in resolvedDays && week.sessions.any { it.dayOfWeek == day }
                        }.toSet()
                        weeks += CalendarWeekProjection(
                            weekId = week.id,
                            weekName = week.name,
                            macroIndex = macroIndex,
                            blockIndex = blockIndex,
                            mesoIndex = mesoIndex,
                            weekIndex = globalWeekIndex,
                            blockId = block.id,
                            blockName = block.name,
                            startDate = weekStart,
                            endDate = weekEnd,
                            outsideProgramDays = outsideDays,
                            trainingDayDates = dayDates,
                            keyDates = marks,
                        )
                        cursor = weekEnd.plusDays(1)
                        globalWeekIndex++
                    }
                }
            }
        }

        val projectedEnd = weeks.lastOrNull()?.endDate
        val manualEnd = parseIsoDate(calendar.manualEndDate) ?: parseIsoDate(plan.targetEndDate)
        val status = endDateStatus(manualEnd, projectedEnd, calendar.manualEndDate ?: plan.targetEndDate)
        return ProgramCalendarProjection(
            enabled = true,
            mode = calendar.mode,
            strictStart = calendar.strictStart,
            activatedByCompetition = calendar.activatedByCompetition,
            startDate = start,
            projectedEndDate = projectedEnd,
            manualEndDate = manualEnd,
            endDateStatus = status,
            weeks = weeks,
        )
    }

    fun materializeWeekDates(program: Program): Program {
        val projection = project(program)
        if (!projection.enabled || projection.weeks.isEmpty()) return program
        val byId = projection.weeks.associateBy { it.weekId }
        return program.copy(
            macrocycles = program.macrocycles.map { macro ->
                macro.copy(
                    blocks = macro.blocks.map { block ->
                        block.copy(
                            mesocycles = block.mesocycles.map { meso ->
                                meso.copy(
                                    weeks = meso.weeks.map { week ->
                                        if (week.isLoopWeek) return@map week
                                        val projected = byId[week.id] ?: return@map week
                                        val preservedTrainingDays = resolveTrainingDays(program, week)
                                        val mergedTrainingDates = projected.trainingDayDates
                                            .filterKeys { day -> day in preservedTrainingDays }
                                            .mapValues { it.value.toString() }
                                        week.copy(
                                            startDate = projected.startDate.toString(),
                                            endDate = projected.endDate.toString(),
                                            trainingDayDates = mergedTrainingDates,
                                        )
                                    }
                                )
                            }
                        )
                    }
                )
            }
        )
    }

    fun scheduleIssueFor(program: Program, weekId: String?, session: Session, actualDate: LocalDate = SystemAppClock.today(java.time.ZoneId.systemDefault())): ScheduleIssue? {
        val projection = project(program)
        if (!projection.enabled) return null
        val planned = weekId?.let { projection.scheduledDateFor(session, it) }
        val actualWeek = projection.weekForDate(actualDate)
        return when {
            actualWeek == null -> ScheduleIssue.OutsideProgram(actualDate, projection.startDate, projection.effectiveEndDate)
            planned != null && planned != actualDate -> ScheduleIssue.WrongDate(planned, actualDate)
            planned == null && weekId != null -> ScheduleIssue.OutsideProgramDay(actualDate)
            else -> null
        }
    }

    fun resolveTrainingDays(program: Program, week: ProgramWeek): Set<Int> {
        val fromPlan = program.resolvedSchedulePlan().trainingDays.filter { it in 1..7 }
        if (fromPlan.isNotEmpty()) return fromPlan.toSet()
        val fromWeek = week.trainingDayDates.keys.filter { it in 1..7 }
        if (fromWeek.isNotEmpty()) return fromWeek.toSet()
        if (program.isSimpleProgram) {
            val fromSessions = week.sessions.mapNotNull { it.dayOfWeek?.takeIf { day -> day in 1..7 } }
            if (fromSessions.isNotEmpty()) return fromSessions.toSet()
            return program.suggestCalendarTrainingDays()
        }
        return (1..7).toSet()
    }

    fun alignToWeekStart(date: LocalDate, weekStartDay: Int?): LocalDate {
        val target = weekStartDay?.coerceIn(1, 7) ?: date.dayOfWeek.value
        val current = date.dayOfWeek.value
        val delta = (current - target + 7) % 7
        return if (delta == 0) date else date.minusDays(delta.toLong())
    }

    private fun trainingDatesFor(
        weekStart: LocalDate,
        weekEnd: LocalDate,
        trainingDays: Set<Int>,
        explicit: Map<Int, String>,
    ): Map<Int, LocalDate> {
        return trainingDays.mapNotNull { day ->
            val explicitDate = parseIsoDate(explicit[day])
            val resolved = explicitDate ?: dateForDay(weekStart, weekEnd, day) ?: return@mapNotNull null
            if (resolved.isBefore(weekStart) || resolved.isAfter(weekEnd)) null else day to resolved
        }.toMap()
    }

    private fun dateForDay(start: LocalDate, end: LocalDate, day: Int): LocalDate? {
        var cursor = start
        while (!cursor.isAfter(end)) {
            if (cursor.dayOfWeek.value == day) return cursor
            cursor = cursor.plusDays(1)
        }
        return null
    }

    private fun keyDateIntersects(keyDate: ProgramKeyDate, weekStart: LocalDate, weekEnd: LocalDate): Boolean {
        val event = parseIsoDate(keyDate.eventDate)
        val start = parseIsoDate(keyDate.startDate) ?: event ?: return false
        val end = parseIsoDate(keyDate.endDate) ?: start
        return !end.isBefore(weekStart) && !start.isAfter(weekEnd)
    }

    private fun endDateStatus(manualEnd: LocalDate?, projectedEnd: LocalDate?, rawManual: String?): ProgramEndDateStatus {
        if (rawManual.isNullOrBlank()) return ProgramEndDateStatus.NONE
        if (manualEnd == null || projectedEnd == null) return ProgramEndDateStatus.INVALID_MANUAL
        val diff = ChronoUnit.DAYS.between(projectedEnd, manualEnd)
        return when {
            diff == 0L -> ProgramEndDateStatus.MATCHES_PROJECTED
            diff < 0L -> ProgramEndDateStatus.BEFORE_PROJECTED
            else -> ProgramEndDateStatus.AFTER_PROJECTED
        }
    }
}

sealed interface ScheduleIssue {
    data class WrongDate(val plannedDate: LocalDate, val actualDate: LocalDate) : ScheduleIssue
    data class OutsideProgram(val actualDate: LocalDate, val startDate: LocalDate?, val endDate: LocalDate?) : ScheduleIssue
    data class OutsideProgramDay(val actualDate: LocalDate) : ScheduleIssue
}

File: ios-native/KPKNFit/KPKNFit/Domain/Training/ProgramCalendarEngine.swift
import Foundation

enum ProgramEndDateStatus {
    case none, matchesProjected, beforeProjected, afterProjected, invalidManual
}

struct CalendarWeekProjection {
    let weekId: String
    let weekName: String
    let macroIndex: Int
    let blockIndex: Int
    let mesoIndex: Int
    let weekIndex: Int
    let blockId: String
    let blockName: String
    let startDate: Date
    let endDate: Date
    let outsideProgramDays: Set<Int>
    let trainingDayDates: [Int: Date]
    let keyDates: [ProgramKeyDate]

    func contains(_ date: Date) -> Bool {
        date >= startDate && date <= endDate
    }
}

struct ProgramCalendarProjection {
    let enabled: Bool
    let mode: ProgramCalendarizationMode?
    let strictStart: Bool
    let activatedByCompetition: Bool
    let startDate: Date?
    let projectedEndDate: Date?
    let manualEndDate: Date?
    let endDateStatus: ProgramEndDateStatus
    let weeks: [CalendarWeekProjection]

    var effectiveEndDate: Date? { manualEndDate ?? projectedEndDate }

    func weekForDate(_ date: Date) -> CalendarWeekProjection? {
        weeks.first { $0.contains(date) }
    }

    func scheduledDateFor(_ session: Session, weekId: String) -> Date? {
        guard let week = weeks.first(where: { $0.weekId == weekId }) else { return nil }
        guard let day = session.dayOfWeek, (1...7).contains(day) else { return week.startDate }
        return week.trainingDayDates[day]
    }
}

enum ProgramCalendarEngine {
    static func parseIsoDate(_ raw: String?) -> Date? {
        guard let raw = raw, !raw.trimmingCharacters(in: .whitespaces).isEmpty else { return nil }
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withFullDate]
        return formatter.date(from: raw)
    }

    static func isCalendarized(_ program: Program) -> Bool {
        guard let calendar = program.calendarization else { return false }
        switch calendar.mode {
        case .advancedCompetition:
            return !program.isSimpleTemporalProgram && !(program.timelineStartDate ?? "").isEmpty
        case .simpleDated:
            return program.isSimpleTemporalProgram && program.simpleProgramKind == .calendarized && !(program.timelineStartDate ?? "").isEmpty
        }
    }

    static func defaultCompetitionCalendarization() -> ProgramCalendarization {
        ProgramCalendarization(mode: .advancedCompetition, strictStart: true, activatedByCompetition: true)
    }

    static func defaultSimpleDatedCalendarization() -> ProgramCalendarization {
        ProgramCalendarization(mode: .simpleDated, strictStart: false, activatedByCompetition: false)
    }

    static func project(_ program: Program) -> ProgramCalendarProjection {
        let calendar = program.calendarization
        let start = parseIsoDate(program.timelineStartDate)
        guard let calendar = calendar, let start = start else {
            return ProgramCalendarProjection(
                enabled: false, mode: calendar?.mode,
                strictStart: calendar?.strictStart == true,
                activatedByCompetition: calendar?.activatedByCompetition == true,
                startDate: start, projectedEndDate: nil,
                manualEndDate: parseIsoDate(calendar?.manualEndDate),
                endDateStatus: .none, weeks: []
            )
        }

        var cursor = start
        var globalWeekIndex = 0
        var weeks: [CalendarWeekProjection] = []

        for (macroIndex, macro) in program.macrocycles.enumerated() {
            var globalMesoIndex = 0
            for (blockIndex, block) in macro.blocks.enumerated() {
                for meso in block.mesocycles {
                    let mesoIndex = globalMesoIndex
                    globalMesoIndex += 1
                    for week in meso.weeks {
                        let weekStart = parseIsoDate(week.startDate) ?? cursor
                        let weekEnd = parseIsoDate(week.endDate) ?? projectedWeekEnd(start: weekStart)
                        let outsideDays = outsideDaysFor()
                        let dayDates = trainingDatesFor(start: weekStart, end: weekEnd, outsideDays: outsideDays, explicit: week.trainingDayDates)
                        let marks = program.keyDates.filter { keyDateIntersects($0, weekStart: weekStart, weekEnd: weekEnd) }
                        weeks.append(CalendarWeekProjection(
                            weekId: week.id, weekName: week.name,
                            macroIndex: macroIndex, blockIndex: blockIndex,
                            mesoIndex: mesoIndex, weekIndex: globalWeekIndex,
                            blockId: block.id, blockName: block.name,
                            startDate: weekStart, endDate: weekEnd,
                            outsideProgramDays: outsideDays, trainingDayDates: dayDates, keyDates: marks
                        ))
                        cursor = Calendar.current.date(byAdding: .day, value: 1, to: weekEnd) ?? weekEnd
                        globalWeekIndex += 1
                    }
                }
            }
        }

        let projectedEnd = weeks.last?.endDate
        let manualEnd = parseIsoDate(calendar.manualEndDate)
        let status = endDateStatus(manualEnd: manualEnd, projectedEnd: projectedEnd, rawManual: calendar.manualEndDate)

        return ProgramCalendarProjection(
            enabled: true, mode: calendar.mode, strictStart: calendar.strictStart,
            activatedByCompetition: calendar.activatedByCompetition,
            startDate: start, projectedEndDate: projectedEnd,
            manualEndDate: manualEnd, endDateStatus: status, weeks: weeks
        )
    }

    static func materializeWeekDates(_ program: Program) -> Program {
        let projection = project(program)
        guard projection.enabled, !projection.weeks.isEmpty else { return program }
        let byId = Dictionary(uniqueKeysWithValues: projection.weeks.map { ($0.weekId, $0) })
        return program.copy(macrocycles: program.macrocycles.map { macro in
            macro.copy(blocks: macro.blocks.map { block in
                block.copy(mesocycles: block.mesocycles.map { meso in
                    meso.copy(weeks: meso.weeks.map { week in
                        guard let projected = byId[week.id] else { return week }
                        return week.copy(
                            startDate: ISO8601DateFormatter().string(from: projected.startDate),
                            endDate: ISO8601DateFormatter().string(from: projected.endDate),
                            trainingDayDates: projected.trainingDayDates.mapValues { ISO8601DateFormatter().string(from: $0) }
                        )
                    })
                })
            })
        })
    }

    static func scheduleIssueFor(program: Program, weekId: String?, session: Session, actualDate: Date = Date()) -> ScheduleIssue? {
        let projection = project(program)
        guard projection.enabled else { return nil }
        let planned = weekId.flatMap { projection.scheduledDateFor(session, weekId: $0) }
        let actualWeek = projection.weekForDate(actualDate)
        if actualWeek == nil {
            return .outsideProgram(actual: actualDate, start: projection.startDate, end: projection.effectiveEndDate)
        }
        if let planned = planned, planned != actualDate {
            return .wrongDate(planned: planned, actual: actualDate)
        }
        if planned == nil && weekId != nil {
            return .outsideProgramDay(actual: actualDate)
        }
        return nil
    }

    private static func projectedWeekEnd(start: Date) -> Date {
        Calendar.current.date(byAdding: .day, value: 6, to: start) ?? start
    }

    private static func outsideDaysFor() -> Set<Int> {
        []
    }

    private static func trainingDatesFor(start: Date, end: Date, outsideDays: Set<Int>, explicit: [Int: String]) -> [Int: Date] {
        var result: [Int: Date] = [:]
        for day in 1...7 where !outsideDays.contains(day) {
            if let explicitDate = parseIsoDate(explicit[day]) {
                if explicitDate >= start && explicitDate <= end { result[day] = explicitDate }
            } else if let resolved = dateForDay(start: start, end: end, day: day) {
                result[day] = resolved
            }
        }
        return result
    }

    private static func dateForDay(start: Date, end: Date, day: Int) -> Date? {
        var cursor = start
        while cursor <= end {
            if Calendar.current.component(.weekday, from: cursor) == day { return cursor }
            cursor = Calendar.current.date(byAdding: .day, value: 1, to: cursor) ?? cursor
        }
        return nil
    }

    private static func keyDateIntersects(_ keyDate: ProgramKeyDate, weekStart: Date, weekEnd: Date) -> Bool {
        guard let event = parseIsoDate(keyDate.eventDate) else { return false }
        let start = parseIsoDate(keyDate.startDate) ?? event
        let end = parseIsoDate(keyDate.endDate) ?? start
        return end >= weekStart && start <= weekEnd
    }

    private static func endDateStatus(manualEnd: Date?, projectedEnd: Date?, rawManual: String?) -> ProgramEndDateStatus {
        guard let raw = rawManual, !raw.isEmpty else { return .none }
        guard let manual = manualEnd, let projected = projectedEnd else { return .invalidManual }
        let diff = Calendar.current.dateComponents([.day], from: projected, to: manual).day ?? 0
        if diff == 0 { return .matchesProjected }
        if diff < 0 { return .beforeProjected }
        return .afterProjected
    }
}

enum ScheduleIssue {
    case wrongDate(planned: Date, actual: Date)
    case outsideProgram(actual: Date, start: Date?, end: Date?)
    case outsideProgramDay(actual: Date)
}
